File: src/exdl_joystick.erl
%%
%%  Copyright (c) 2001 Dan Gudmundsson
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%% 
%%     $Id: exdl_joystick.erl,v 1.2 2005/06/29 14:45:39 dgud Exp $
%%

-module(exdl_joystick).

-include("exdl.hrl").

%% 1- EXPORT

%-compile(export_all).

-export( [ numJoysticks/0
         , name/1
         , open/1
         , opened/1
         , index/1
         , numAxes/1
         , numBalls/1
         , numHats/1
         , numButtons/1
         , update/0
         , eventState/1
         , getAxis/2
         , getHat/2
         , getBall/2
         , getButton/2
         , close/1
         ]
       ).

%% 2- INCLUDE

-include("exdl_util.hrl").

-include("exdl_joystick.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

%% 4- FUNCTIONS

%% Func:       numJoysticks
%% Args:       none
%% Returns:    Number of joysticks
%% C-API func: int SDL_NumJoysticks(void)
%% Desc:       Count the number of joysticks attached to the system 

numJoysticks()
 -> case
      call
      ( ?SDL_NumJoysticks
      , [
        ]
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       name
%% Args:       Deviceindex
%% Returns:    Name (List)
%% C-API func: const char* SDL_JoystickName(int device_index)
%% Desc:       Get the implementation dependent name of a joystick.
%%             This can be called before any joysticks are opened.
%%             If no name can be found, this function returns NULL.

name(DeviceIndex)
 -> Name
    = call
      ( ?SDL_JoystickName
      , <<DeviceIndex:?Int
        >>
      )
  , binary_to_list
    ( Name
    )
.

%% Func:       open
%% Args:       DeviceIndex
%% Returns:    Joystick or exits
%% C-API func: SDL_Joystick* SDL_JoystickOpen(int device_index);
%% Desc:       Open a joystick for use - the index passed as an argument refers to
%%             the N'th joystick on the system.  This index is the value which will
%%             identify this joystick in future joystick events.
%%             This function returns a joystick identifier, or NULL if an error occurred.

open(DeviceIndex)
 -> <<ID:?_PTR
    >>
    = call
      ( ?SDL_JoystickOpen
      , [ DeviceIndex
        ]
      )
  , ID
.

%% Func:       opened
%% Args:       DeviceIndex
%% Returns:    true or false
%% C-API func: int SDL_JoystickOpened(int device_index);
%% Desc:       Returns true if the joystick has been opened, or false if it has not. 

opened(DeviceIndex)
 -> <<Bool:8
    >>
    = call
      ( ?SDL_JoystickOpened
      , <<DeviceIndex:?Int
        >>
      )
  , Bool == 1
.

%% Func:       index
%% Args:       Joystick
%% Returns:    index
%% C-API func: int SDL_JoystickIndex(SDL_Joystick* joystick);
%% Desc:       Get the device index of an opened joystick.

index(Joystick)
 -> <<Index:8
    >>
    = call
      ( ?SDL_JoystickIndex
      , <<Joystick:?_PTR
        >>
      )
  , Index
.

%% Func:       numAxes
%% Args:       Joystick
%% Returns:    no of axes
%% C-API func: int SDL_JoystickNumAxes(SDL_Joystick* joystick);
%% Desc:       Get the number of general axis controls on a joystick

numAxes(Joystick)
 -> <<Axes:8
    >>
    = call
      ( ?SDL_JoystickNumAxes
      , <<Joystick:?_PTR
        >>
      )
  , Axes
.

%% Func:       numBalls
%% Args:       Joystick
%% Returns:    no of balls
%% C-API func: int SDL_JoystickNumBalls(SDL_Joystick* joystick);
%% Desc:       Get the number of trackballs on a joystick,
%%             joystick trackballs have only relative motion events associated
%%             with them and their state cannot be polled.

numBalls(Joystick)
 -> <<Balls:8
    >>
    = call
      ( ?SDL_JoystickNumBalls
      , <<Joystick:?_PTR
        >>
      )
  , Balls
.

%% Func:       numHats
%% Args:       Joystick
%% Returns:    no of Hats
%% C-API func: int SDL_JoystickNumHats(SDL_Joystick* joystick);
%% Desc:       Get the number of POV hats on a joystick

numHats(Joystick)
 -> <<Hats:8
    >>
    = call
      ( ?SDL_JoystickNumHats
      , <<Joystick:?_PTR
        >>
      )
  , Hats
.

%% Func:       numButtons
%% Args:       Joystick
%% Returns:    no of Buttons
%% C-API func: int SDL_JoystickNumButtons(SDL_Joystick* joystick);
%% Desc:       Get the number of buttons on a joystick

numButtons(Joystick)
 -> <<Buttons:8
    >>
    = call
      ( ?SDL_JoystickNumButtons
      , <<Joystick:?_PTR
        >>
      )
  , Buttons
.

%% Func:       update
%% Args:       none
%% Returns:    none
%% C-API func: void SDL_JoystickUpdate(void);
%% Desc:       Update the current state of the open joysticks.
%%             This is called automatically by the event loop if any joystick
%%             events are enabled.

update()
 -> cast
    ( ?SDL_JoystickUpdate
    , [
      ]
    )
.

%% Func:       getAxis
%% Args:       Joystick, Axis
%% Returns:    State (Int range: -32768 to 32767)
%% C-API func: Sint16 SDL_JoystickGetAxis(SDL_Joystick* joystick, int axis);
%% Desc:       Get the current state of an axis control on a joystick
%%             The state is a value ranging from -32768 to 32767.
%%             The axis indices start at index 0.

getAxis(Joystick, Axis)
 -> <<State:32
    >>
    = call
      ( ?SDL_JoystickGetAxis
      , <<Joystick:32
        , Axis:8
        >>
      )
  , State
.

%% Func:       getHat
%% Args:       Joystick, Hat
%% Returns:    HatState 
%% C-API func: Uint8 SDL_JoystickGetHat(SDL_Joystick* joystick, int hat);
%% Desc:       Get the current state of a POV hat on a joystick
%%             The return value is one of the following positions in sdl_joystick.hrl
%%             The hat indices start at index 0.

getHat(Joystick, Hat)
 -> <<State:8
    >>
    = call
      ( ?SDL_JoystickGetHat
      , <<Joystick:32
        , Hat:8
        >>
      )
  , State
.

%% Func:       getBall
%% Args:       Joystick, Ball
%% Returns:    {DX, DY} or badarg
%% C-API func: int SDL_JoystickGetBall(SDL_Joystick* joystick, int ball, int* dx, int* dy);
%% Desc:       Get the ball axis change since the last poll
%%             This returns 0, or -1 if you passed it invalid parameters.
%%             The ball indices start at index 0.

getBall(Joystick, Ball)
 -> case
      call
      ( ?SDL_JoystickGetBall
      , <<Joystick:32
        , Ball:16
        >>
      )
    of
      <<Dx:32
      , Dy:32
      >>
        -> {Dx, Dy}
    ; Res
        -> {badarg, Res}
    end
.

%% Func:       getButton
%% Args:       Joystick, Button
%% Returns:    State
%% C-API func: Uint8 SDL_JoystickGetButton(SDL_Joystick* joystick, int button);
%% Desc:       Get the current state of a button on a joystick
%%             The button indices start at index 0.

getButton(Joystick, Button)
 -> <<State:8
    >>
    = call
      ( ?SDL_JoystickGetButton
      , <<Joystick:32
        , Button:8
        >>
      )
  , State
.

%% Func:       close
%% Args:       Joystick
%% Returns:    none
%% C-API func: void SDL_JoystickClose(SDL_Joystick* joystick);
%% Desc:       Close a joystick previously opened with SDL_JoystickOpen()

close(Joystick)
 -> cast
    ( ?SDL_JoystickClose
    , <<Joystick:32
      >>
    )
.

File: priv/src/exdl_joystick.c
/*
 *  Copyright (c) 2001 Dan Gudmundsson
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_joystick.c,v 1.2 2005/06/29 14:45:38 dgud Exp $
 */

#include "exdl.h"

#include <sdl\SDL_joystick.h>

EXDL_API(ejoy_numJoysticks)
{
  /* int SDL_NumJoysticks(void) */

  /* exp : Counts the number of joysticks attached to the system. */
  /* inp :  */
  /* ret : the number of attached joysticks */

  CATCH_RESULT(int)

    SDL_NumJoysticks
    (
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_name)
{
  /* const char* SDL_JoystickName(int device_index) */

  /* exp : Get the implementation dependent name of joystick. */
  /* inp : index = the N'th joystick on the system. */
  /* ret : a char pointer to the joystick name. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, device_index)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    SDL_JoystickName
    ( STACK_GET_DATA(device_index)
    );

  THROW_STRING()
}

EXDL_API(ejoy_open)
{
  /* SDL_Joystick* SDL_JoystickOpen(int device_index) */

  /* exp : Opens a joystick for use within SDL. */
  /*     > A joystick must be opened before it game be used */
  /* inp : index = the N'th joystick in the system. */
  /* ret : a SDL_Joystick structure on success. NULL on failure. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, device_index)
  STACK_FMT_END()

  CATCH_RESULT(SDL_Joystick*)

    SDL_JoystickOpen
    ( STACK_GET_DATA(device_index)
    );

  THROW_POINTER()
}

EXDL_API(ejoy_opened)
{
  /* int SDL_JoystickOpened(int device_index) */

  /* exp : Determines whether a joystick has already been opened within the application */
  /* inp : index = the N'th joystick on the system. */
  /* ret : 1 if the joystick has been opened, or 0 if it has not. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, device_index)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickOpened
    ( STACK_GET_DATA(device_index)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_index)
{
  /* int SDL_JoystickIndex(SDL_Joystick* joystick) */

  /* exp : Returns the index of a given SDL_Joystick structure. */
  /* inp : joystick = SDL_Joystick structure. */
  /* ret : Index number of the joystick. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick*, joystick)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickIndex
    ( STACK_GET_DATA(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numAxes)
{
  /* int SDL_JoystickNumAxes(SDL_Joystick* joystick) */

  /* exp : Return the number of axes available from a previously opened SDL_Joystick. */
  /* inp : joystick = SDL_Joystick structure. */
  /* ret : Number of axes. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick*, joystick)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumAxes
    ( STACK_GET_DATA(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numBalls)
{
  /* int SDL_JoystickNumBalls(SDL_Joystick* joystick) */

  /* exp : Return the number of trackballs available from a previously opened SDL_Joystick. */
  /* inp : joystick = SDL_Joystick structure. */
  /* ret : Number of trackballs. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick*, joystick)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumBalls
    ( STACK_GET_DATA(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numHats)
{
  /* int SDL_JoystickNumHats(SDL_Joystick* joystick) */

  /* exp : Return the number of hats available from a previously opened SDL_Joystick. */
  /* inp : joystick = SDL_Joystick structure. */
  /* ret : Number of hats. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick*, joystick)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumHats
    ( STACK_GET_DATA(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numButtons)
{
  /* int SDL_JoystickNumButtons(SDL_Joystick* joystick) */

  /* exp : Return the number of buttons available from a previously opened SDL_Joystick. */
  /* inp : joystick = SDL_Joystick structure. */
  /* ret : Number of buttons. */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick*, joystick)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumButtons
    ( STACK_GET_DATA(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_update)
{
  /* void SDL_JoystickUpdate(void) */

  /* exp : Updates the state(position, buttons, etc.) of all open joysticks. */
  /*       If joystick events have been enabled with SDL_JoystickEventState */
  /*       then this is called automatically in the event loop. */
  /* inp :  */
  /* ret :  */

  SDL_JoystickUpdate
  (
  );
}

EXDL_API(ejoy_eventState)
{
  /* int SDL_JoystickEventState(int state) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, state)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickEventState
    ( STACK_GET_DATA(state)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_getAxis)
{
  /* Sint16 SDL_JoystickGetAxis(SDL_Joystick* joystick, int axis) */

  /* exp : SDL_JoystickGetAxis returns the current state of the given axis on the */
  /*       given joystick. On most modern joysticks the X axis is usually */
  /*       represented by axis 0 and the Y axis by axis 1. The value returned by */
  /*       SDL_JoystickGetAxis is a signed integer (-32768 to 32768) representing */
  /*       the current position of the axis, it maybe necessary to impose certain */
  /*       tolerances on these values to account for jitter. It is worth noting */
  /*       that some joysticks use axes 2 and 3 for extra buttons.*/
  /* inp :  */
  /* ret : a 16-bit signed integer representing the current position of the axis. */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           axis)
  STACK_FMT_END()

  CATCH_RESULT(Sint16)

    SDL_JoystickGetAxis
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(axis)
    );

  THROW_RESULT(Sint16)
}

EXDL_API(ejoy_getHat)
{
  /* Uint8 SDL_JoystickGetHat(SDL_Joystick* joystick, int hat) */

  /* exp : SDL_JoystickGetHat returns the current state of the given hat on the given joystick. */
  /* inp :  */
  /* ret : The current state is returned as a Uint8 which is defined as an OR'd */
  /*       combination of one or more of the following */
  /*       SDL_HAT_CENTERED | SDL_HAT_UP | SDL_HAT_RIGHT | SDL_HAT_DOWN | SDL_HAT_LEFT */
  /*       SDL_HAT_RIGHTUP | SDL_HAT_RIGHTDOWN | SDL_HAT_LEFTUP | SDL_HAT_LEFTDOWN */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           hat)
  STACK_FMT_END()

  CATCH_RESULT(Uint8)

    SDL_JoystickGetHat
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(hat)
    );

  THROW_RESULT(Uint8)
}

EXDL_API(ejoy_getBall)
{
  /* int SDL_JoystickGetBall(SDL_Joystick* joystick, int ball, int* dx, int* dy) */

  /* exp : Get the ball axis change. */
  /*     > Trackballs can only return relative motion since the last call to */
  /*       SDL_JoystickGetBall, these motion deltas a placed into dx and dy. */
  /* inp :  */
  /* ret : 0 on success or -1 on failure */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           ball)
  STACK_FMT_END()

  FRAME_FMT_BEGIN()
    FRAME_FMT_DATA(int, status)
    FRAME_FMT_DATA(int, dx)
    FRAME_FMT_DATA(int, dy)
  FRAME_FMT_END()

  FRAME_SET_DATA(status) =
//  CATCH_RESULT(int)

    SDL_JoystickGetBall
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(ball)
    , FRAME_GET_ADDR(dx)
    , FRAME_GET_ADDR(dy)
    );

  THROW_FRAME()
}

EXDL_API(ejoy_getButton)
{
  /* Uint8 SDL_JoystickGetButton(SDL_Joystick* joystick, int button) */

  /* exp : SDL_JoystickGetButton returns the current state of the given */
  /*       button on the given joystick. */
  /* inp :  */
  /* ret : 1 if the button is pressed. Otherwise, 0. */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           button)
  STACK_FMT_END()

  CATCH_RESULT(Uint8)

    SDL_JoystickGetButton
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(button)
    );

  THROW_RESULT(Uint8)
}

EXDL_API(ejoy_close)
{
  /* void SDL_JoystickClose(SDL_Joystick* joystick) */

  /* exp : Close a joystick that was previously opened with SDL_JoystickOpen. */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick*, joystick)
  STACK_FMT_END()

    SDL_JoystickClose
    ( STACK_GET_DATA(joystick)
    );
}

